go-controller/pkg/crd/types/networkselector.go

```diff
@@ -0,0 +1,88 @@
+package types
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NetworkSelectors selects multiple sets of networks.
+// +kubebuilder:validation:MinItems=1
+// +kubebuilder:validation:MaxItems=5
+// +patchMergeKey=networkSelectionType
+// +patchStrategy=merge
+// +listType=map
+// +listMapKey=networkSelectionType
+type NetworkSelectors []NetworkSelector
+
+// NetworkSelector selects a set of networks.
+// +kubebuilder:validation:XValidation:rule="!has(self.clusterUserDefinedNetworkSelector) || has(self.networkSelectionType) && self.networkSelectionType == 'ClusterUserDefinedNetworks'",message="Can only specify clusterUserDefinedNetworkSelector when networkSelectionType is ClusterUserDefinedNetworks"
+// +kubebuilder:validation:XValidation:rule="!has(self.primaryUserDefinedNetworkSelector) || has(self.networkSelectionType) && self.networkSelectionType == 'PrimaryUserDefinedNetworks'",message="Can only specify primaryUserDefinedNetworkSelector when networkSelectionType is PrimaryUserDefinedNetworks"
+// +kubebuilder:validation:XValidation:rule="!has(self.secondaryUserDefinedNetworkSelector) || has(self.networkSelectionType) && self.networkSelectionType == 'SecondaryUserDefinedNetworks'",message="Can only specify secondaryUserDefinedNetworkSelector when networkSelectionType is SecondaryUserDefinedNetworks"
+// +kubebuilder:validation:XValidation:rule="!has(self.networkAttachmentDefinitionSelector) || has(self.networkSelectionType) && self.networkSelectionType == 'NetworkAttachmentDefinitions'",message="Can only specify networkAttachmentDefinitionSelector when networkSelectionType is NetworkAttachmentDefinitions"
+type NetworkSelector struct {
+	// networkSelectionType determines the type of networks selected.
+	// +kubebuilder:validation:Required
+	NetworkSelectionType NetworkSelectionType `json:"networkSelectionType"`
```

Comment:
Have you considered using a `unionDiscriminator`? I think this could help remove the CEL validation above.

Reply:
Can you give me a reference?

Reply:
https://github.com/ovn-kubernetes/ovn-kubernetes/blob/96b8798640db4314dd915716b9f8546dd88ebec2/go-controller/pkg/crd/userdefinednetwork/v1/cudn.go#L40-L61

Reply:
Thanks. So I did think somethign like this might be available but I could find doc in  I can't find doc
https://book.kubebuilder.io/introduction?search=unionDiscriminator
Is this openshift specific? Is there doc?

Reply:
I don't think it is a kubebuilder feature, here is a KEP I found but it doesn't exactly match what we used for UDN
https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1027-api-unions/README.md

Seems we used the OCP verison:
https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions

Reply:
here is an upstream example that seems the same to what we are using: https://github.com/kubernetes/kubernetes/blob/ab3e83f73424a18f298a0050440af92d2d7c4720/pkg/apis/core/types.go#L4020-L4031

Reply:
I can throw it in.
But precisely what validation do you think it replaces? Can we be 100% sure of what it is doing to remove that validation?

Reply:
No changes at all in the schema with or without it. I will just throw it in to be aligned with the UDN spec.